### PR TITLE
infrastructure for UI/integration tests that read the logfile

### PR DESCRIPTION
**Pull request by individual-it**
## Description
this is the infrastructure for UI/integration tests that need to read/analyse the log file

## Motivation and Context
tests might want to read the log file to see if a particular event was logged and what was logged

## How Has This Been Tested?
wrote tests for firewall app that test the log output

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: tests/TestHelpers/LoggingHelper.php [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+		if ($result["code"] != 0) {
+			throw new \Exception(
+				"could not get log backend " . $result ["stdOut"] . " " . $result ["stdErr"]
+			);
+		}
+		$pregResult = preg_match(
+			"/Enabled logging backend:\s(\w+)\n/",
+			$result ["stdOut"], $matches
+		);
+		if (!$pregResult) {
```
Comment (phil-davis) [inline tests/TestHelpers/LoggingHelper.php]: The line above could be put inside the if() like you did in getLogLevel - just to be consistent.

File: tests/TestHelpers/LoggingHelper.php [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+		$result = SetupHelper::runOcc(["log:manage"], $ocPath);
+		if ($result["code"] != 0) {
+			throw new \Exception(
+				"could not get log timezone " . $result ["stdOut"] . " " . $result ["stdErr"]
+			);
+		}
+		$pregResult = preg_match(
+			"/Log timezone:\s(\w+)/", $result ["stdOut"], $matches
+		);
+		if (!$pregResult) {
```
Comment (phil-davis) [inline tests/TestHelpers/LoggingHelper.php]: And here - $pregResult can be eliminated.

File: tests/ui/features/bootstrap/Logging.php [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+					if (isset($logEntries [$attribute]) &&
+						  (
+							$logEntryExpectedNotToExist [$attribute] === "" ||
+							$logEntryExpectedNotToExist [$attribute] === $logEntries [$attribute]
+						  )
+						) {
+						$match = true;
+					} else {
+						$match = false;
+						break;
```
Comment (phil-davis) [inline tests/ui/features/bootstrap/Logging.php]: I think this break needs to be at the end of the "if" section???
To pass the overall test we want no matches in the log file. So as soon as we find a match, we break.
If there is no match, then we keep looping to see if we can find any match.
Maybe can remove the else completely, and start with ```$match = false;``` 
And does it need to ``break 2``` out of 2 for loops to get out to the PHPUnit_Framework_Assert ?

File: tests/ui/features/bootstrap/Logging.php [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+	public function setUpScenarioLogging(BeforeScenarioScope $scope)
+	{
+		$this->oldLogLevel = LoggingHelper::getLogLevel($this->ocPath);
+		$this->oldLogBackend = LoggingHelper::getLogBackend($this->ocPath);
+		$this->oldLogTimezone = LoggingHelper::getLogTimezone($this->ocPath);
+	}
+
+	/** @AfterScenario */
+	public function tearDownScenarioLogging(AfterScenarioScope $scope)
+	{
```
Comment (phil-davis) [inline tests/ui/features/bootstrap/Logging.php]: should this be made even fancier?
the code could remember if each of log level, backend or timezone have actually been changed. then only bother to return the settings if they were changed.

**Review by phil-davis — COMMENTED** [review]
See individual comments for discussion.

File: tests/TestHelpers/LoggingHelper.php (hunk 1 above)
Comment (individual-it) [inline tests/TestHelpers/LoggingHelper.php]: you are talking about doing like this: `if (!preg_match("/Enabled logging backend:\s(\w+)\n/", $result ["stdOut"], $matches)) {`
is the same functionality but the line gets pretty long and I find it confusing if the if line is broken

Comment (phil-davis) [inline tests/TestHelpers/LoggingHelper.php]: OK - I guess the other place had a shorter regex.

File: tests/ui/features/bootstrap/Logging.php (hunk 3 above)
Comment (individual-it) [inline tests/ui/features/bootstrap/Logging.php]: to pass the test we want that no line in the log file matches all given attributes. In every line there has to be at least one attribute that is different to the attribute given to the test.
e.g. 
logfile:
`line 0: user=>user1, app=> firewall, method=> GET, message=> xxx
line 1: user=>user2, app=> firewall, method=> GET, message=> yyyy
`
table given to the test:
`
|user  |app     |method|message             |
|user1|firewall|GET   |yyyy|
`
in that case the test would pass, because line 0 of the logfile does not match the message attribute and line 1 does not match the users attribute

basically it says: "check if there is no line that is exactly like that given table"

would be great if you could advise on better wording for the rule and the comment

Comment (phil-davis) [inline tests/ui/features/bootstrap/Logging.php]: OK
When some attribute does not match, that means that this log entry is not a match, so break out of bothering to check any more attributes for this entry - that is the break here, which gets out of:
```
foreach (array_keys($logEntryExpectedNotToExist) as $attribute)
```
early.
That will then start on the next $logEntryExpectedNotToExist.

But at this point, if $match = true then we have found a log line that matches one of the logEntriesExpectedNotToExist - so I think we are done? We can (should) break out of the next outer foreach. Otherwise our "memory" of $match could be different for the next loop.

Actually, I think PHPUnit_Framework_Assert::assertFalse can just be moved 1 level in, so it is checked every time the inner loop is finished. If that assertFalse fails, then it will throw all the way out of however many loops are nested.

Comment (phil-davis) [inline tests/ui/features/bootstrap/Logging.php]: This is the type of thing that needs test of the testing infrastructure, to ensure it will really report pass/fail correctly under different conditions - that feels like an infinite recursion coming along.

File: tests/ui/features/bootstrap/Logging.php (hunk 4 above)
Comment (individual-it) [inline tests/ui/features/bootstrap/Logging.php]: made it fancier https://github.com/owncloud/core/pull/28458/commits/ff758413bd1b90c1c2c5ce23ac5cdcbab1ea92d0

File: tests/ui/features/bootstrap/Logging.php [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+	public function theLastLinesOfTheLogFileShouldContainEntriesWithTheseAttributes(
+		TableNode $expectedLogEntries
+	) {
+		//-1 because getRows gives also the header
+		$linesToRead = count($expectedLogEntries->getRows())-1;
+		$logLines = LoggingHelper::tailFile(
+			LoggingHelper::getLogFilePath($this->ocPath),
+			$linesToRead
+		);
+		$lineNo=0;
```
Comment (phil-davis) [inline tests/ui/features/bootstrap/Logging.php]: $lines = 0;

File: tests/ui/features/bootstrap/Logging.php [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+	 * e.g.
+	 * |user|app|method|message|
+	 * @Then the last lines of the log file should contain log-entries with these attributes:
+	 * @return void
+	 */
+	public function theLastLinesOfTheLogFileShouldContainEntriesWithTheseAttributes(
+		TableNode $expectedLogEntries
+	) {
+		//-1 because getRows gives also the header
+		$linesToRead = count($expectedLogEntries->getRows())-1;
```
Comment (phil-davis) [inline tests/ui/features/bootstrap/Logging.php]: count() - 1;

**Review by phil-davis — COMMENTED** [review]
Just some = spacing

**Review by phil-davis — APPROVED** [review]
Looks good now.

**Comment by phil-davis** [comment]
Backport to stable10 #28500
